### PR TITLE
Add standards docs describing SPIFFE and X.509 SVID

### PR DESCRIPTION
**Pull request by evan2645**
This PR checks in the first pass at a more formalized definition of SPIFFE and X.509-based SVIDs. There is (obviously) a lot more detail here than was previously defined. All of the previously defined semantics have been preserved, however we have added quite a bit on top. There are almost certainly errors and oversights. Please review carefully. This PR serves as a discussion ground - additional commits will be pushed here as issues are raised.

### CODE REVIEW

**Comment by evan2645** [comment]
As a note, there will be a future PR which checks in a document describing the Workload API. We are also considering a document which will describe an mTLS profile for SPIFFE.

File: standards/SPIFFE.md
Comment (cyli) [inline standards/SPIFFE.md]: Should "Trust" be capitalized, as it's the start of a new sentence?  Otherwise should it be quoted with `` or otherwise marked as a proper noun or term?  Also, should "Unlike public DNS` be its own sentence, since it expresses a new idea?

File: standards/SPIFFE.md
Comment (cyli) [inline standards/SPIFFE.md]: > that is it can be assumed that the infrastructure that assigned identities under that domain has had trust pre-established prior to issuing those identities

Would it make sense to clarify "trust pre-established"?  It's a little vague here, although it's it's clearer in the SVID section (signing authority for trust domain)

File: standards/SPIFFE.md
Comment (cyli) [inline standards/SPIFFE.md]: Non-blocking question:  This says that the SPIFFE ID of the signing authority SHOULD reside in the trust domain, but the x509 SVID specification says that that CA's SPIFFE ID *MUST* reside in the trust domain.  Do you anticipate some other identity for which the SPIFFE ID cannot reside within that trust domain?

File: standards/X509-SVID.md
Comment (cyli) [inline standards/X509-SVID.md]: Should this require that signing certificates not have the extended key usages for server authentication or client authentication?  Similarly below when key usages for signing certificates are described.  In general CAs shouldn't have extended key usage info, although it's possible they may.

File: standards/X509-SVID.md
Comment (cyli) [inline standards/X509-SVID.md]: Nitpick:  section 4.2.1.12 (missing the dot 1 between the 2 and 12)

File: standards/X509-SVID.md
Comment (cyli) [inline standards/X509-SVID.md]: I think the above:
> Specifically, the SPIFFE ID of the signing certificate MUST be equal to the leaf certificate’s SPIFFE trust domain

Still must be checked, even with URI name constraints.  The name constraints I think only apply to certificates signed by the CA with that name constraint.  If the CA must also have an SPIFFE ID that equals the trust domain, I think that needs to be a separate check no matter what.

Also, do we need to verify that the URI scheme is `spiffe://`, since the URI name constraint I think isn't supposed to have a scheme?  (at least not in openssl, although I didn't see anything in the RFC saying whether or not a scheme should be part of the name constraint)

File: standards/X509-SVID.md
Comment (cyli) [inline standards/X509-SVID.md]: Similarly, does the validator need to verify the SPIFFE ID + domain for each intermediate all the way up the chain to the root?

**Review by cyli — COMMENTED** [review]
A couple questions on the x509 SVID format, but other than that this looks great and has a lot more info than the previous doc!

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Agreed - this bit was copy/pasted from the original SVID specification. IMO, "trust root of a system" is fairly clear, perhaps we could fix this up by simply removing the snippet you quoted?

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Oops, this is a copy-pasta error! I'll address it in a follow-on commit. Thank you :)

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: That's a good question :)... I'm not quite sure, TBH. In the context of X.509, I had a hard time envisioning scenarios in which the SPIFFE ID _wouldn't_ live in the trust domain (else it will break validation), which is why I defined it as MUST under that spec. At the same time, I was unsure about other potential/future SVID formats, so left the higher level spec as SHOULD in order to give some breathing room. To some extent, this is also an artifact of me having written the SPIFFE spec first, and later seeing the "MUST" requirement when writing the X.509 spec, in particular...

It feels like the breathing room is a good thing, especially since we can upgrade the requirement in format-specific specs... that said, I can see how it may lead to confusion.

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: Oops, will fix, thanks!

File: standards/SPIFFE.md
Comment (ajessup) [inline standards/SPIFFE.md]: (Non blocking) Consider giving the three components names, and introducing those names in this section. Helps signpost folks for later.

File: standards/SPIFFE.md
Comment (ajessup) [inline standards/SPIFFE.md]: (Non blocking) Consider removing spurious line break here.

File: standards/SPIFFE.md
Comment (ajessup) [inline standards/SPIFFE.md]: (Non-blocking) Consider introducing names for the typed SVIDs. For example `SVID-X.509` or `SVID-JWT`.

File: standards/SPIFFE.md
Comment (ajessup) [inline standards/SPIFFE.md]: (Non-blocking) Consider linking to the SPIFFE Workload API specification

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: Hmmm good point. Perhaps this can be addressed with two changes:
1. An addition to this last paragraph, referencing section 4.3 and 4.4 for restrictions supporting this point
1. An addition to section 4.4 stating that signing certs MUST NOT define an extended key usage extension

Do you think that will sufficiently address it?

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: No, this validation must only occur between the leaf cert and its immediate authority. One of the use cases we wanted to support was hierarchical trust domains, in which one trust domain (i.e. example.com) may issue a signing certificate for a "child" trust domain (i.e. dev.example.com). Name constraints support this case nicely with leading wildcards.

In supporting this case, if one were to walk all the way up the chain, they may encounter a signing certificate which has signed a signing certificate for a different trust domain, breaking the validation asserted here.

I'd be interested to hear if this breaks security in some way. Assuming it's safe to proceed in this fashion, we can put a small addition to this section explicitly calling out the fact that name validation does not go all the way up the chain

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: > Still must be checked, even with URI name constraints. The name constraints I think only apply to certificates signed by the CA with that name constraint. If the CA must also have an SPIFFE ID that equals the trust domain, I think that needs to be a separate check no matter what.

Hmm I see your point... Perhaps "loosen" is the wrong word. Propose replacing it with "update" and "in order to better leverage name constraint validation." ?

> Also, do we need to verify that the URI scheme is spiffe://, since the URI name constraint I think isn't supposed to have a scheme? (at least not in openssl, although I didn't see anything in the RFC saying whether or not a scheme should be part of the name constraint)

Oh yea... 5280 does actually call out that URI name constraint excludes the scheme, however it does so in a way that could be more clear ("For URIs, the constraint applies to the host part of the name"). To address this, I propose amending paragraph two of this section to add "The validator must also ensure that the scheme of the SPIFFE ID is set to `spiffe://`". Does this sufficiently address the concern?

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Oops, will fix

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Will do, as soon as it's written :) I think there are several places that could do with the addition of some links, but have been putting them off until things are a little more stable

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: I'm having a little trouble unpacking this. I felt like I defined names for them at the end of this paragraph (and followed those names in the section titles) but it seems that it's not obvious enough. Could you provide a suggested edit(s) to help me better understand how you're thinking about this?

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Do you think the names should be introduced here, or in the respective format-specific docs? One consideration is that we probably don't want to have to rev this spec in order to add additional SVID types.

I landed on `X.509-SVID` as a name (see spec filename). I can call this out more clearly in the Introduction section of that document. I ended up putting "X.509" before "SVID" because the title of the document strongly echoes that ordering, and I couldn't come up with a clean title for the opposite ordering... Thoughts?

File: standards/SPIFFE.md
Comment (cyli) [reply standards/SPIFFE.md]: Sure, that works.

File: standards/X509-SVID.md
Comment (cyli) [reply standards/X509-SVID.md]: I think so, yes.

File: standards/X509-SVID.md
Comment (cyli) [reply standards/X509-SVID.md]: > Hmm I see your point... Perhaps "loosen" is the wrong word. Propose replacing it with "update" and "in order to better leverage name constraint validation." ?

Makes sense, thanks!

> To address this, I propose amending paragraph two of this section to add "The validator must also ensure that the scheme of the SPIFFE ID is set to spiffe://". Does this sufficiently address the concern?

Yes, that'd work.

File: standards/SPIFFE.md
Comment (cyli) [reply standards/SPIFFE.md]: Yep, less restrictions at the higher level make sense.  Would it be clearer to add something about how specific types of SVIDs can have tighter restrictions?

File: standards/X509-SVID.md
Comment (cyli) [reply standards/X509-SVID.md]: Ok, so to be clear, we want to support the case where we might have:

1. CA1: no SPIFFE ID, no name constraint, self-signed
2. CA2: no SPIFFE ID, URI name constraint: .example.com, example.com, signed by CA1
3. CA3: spiffe://test.example.com, URI name constraint: test.example.com, signed by CA2
4. Leaf: spiffe://test.example.com/path, signed by CA3

In which case, if the leaf were also an mTLS cert, I totally understand why we don't want to validate all the way up the chain.  But for SPIFFE purposes, would CA1, CA2, or CA3 be considered the trust root of the domain?  Would it be CA3 because it's the only one with a SPIFFE ID?  Would it be CA1 because that's the root of the trust chain for mTLS?  Is there a difference?

File: standards/SPIFFE.md
Comment (ajessup) [reply standards/SPIFFE.md]: Ah, you do indeed - and that looks fine. I skim-read this a little too aggressively. Please disregard.

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: I can try to re-work this paragraph a bit to squeeze something in to that effect. I'd be curious to hear some more opinions on this as well (i.e., perhaps we just make it a MUST here and call it a day)

File: standards/SPIFFE.md
Comment (ajessup) [reply standards/SPIFFE.md]: I buy the argument that we don't want to iterate all the SVID types here, though it might be worth calling out and linking to `X.509-SVID` as an example so people can more clearly understand that the general SVID spec need describe anything that's document specific. I don't feel strongly about this though.

`X.509-SVID` sounds great to me, I think I prefer it to the other ordering.

**Review by ajessup — APPROVED** [review]
LGTM, modulo non-blocking comments. Suggest getting approval from a couple of others before merging.

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: Yes, your example is the case we want to support. As for which cert serves as the root for SPIFFE purposes, that's a good question... At the moment, it's undefined. I suppose it's up to the operator as to where to root the trust (since the operator can presumably compose the CA bundle which it gives to workloads). I don't _think_ there's a difference...

Do you think we need to be opinionated about where to root trust in an X.509 SVID chain? As you pointed out, it's hard to say since upstream CAs may not be SPIFFE-aware.

**Comment by evan2645** [comment]
Some additional key-related questions were raised today... How do folks feel about adding a section addressing acceptable key types and lengths? Suggestions for a minimum bar would also be good. As a starting point, it seems 2048-bit RSA should be supported, as well as ECDSA with whatever the key size equivalent is there.

File: standards/SPIFFE.md
Comment (smawson) [inline standards/SPIFFE.md]: It seems strange to mention the workload API in this doc, but then define it separately.

Is the intent that there are two main specifications?
1) ID + SVID
2) Workload API

If so maybe an overview doc that then links to the other two docs, and those can focus just on their parts? I think you've managed to make those completely orthogonal, which is awesome.

File: standards/X509-SVID.md
Comment (smawson) [inline standards/X509-SVID.md]: Is this only one URI SAN, or only one spiffe URI SAN?

What if I wanted to have an X.509 cert that includes both the identity and the naming information?

Or what about systems that have multiple identities?  Is the assumption that they will use SNI to choose a cert corresponding to the requested entity?

File: standards/X509-SVID.md
Comment (smawson) [inline standards/X509-SVID.md]: what is a SPIFFE service?

**Review by smawson — COMMENTED** [review]
Some peanut gallery comments.

File: standards/SPIFFE.md
Comment (evan2645) [reply standards/SPIFFE.md]: Hmm fair point... Originally I had planned for this doc to serve as the overview doc, but on your comment I realize it has become larger than I had envisioned. I will work on a new (much shorter) doc which speaks to SPIFFE at a very high level, remove section 4 from this spec, and re-work the abstract and introduction.

I imagine we want the high level doc to retain the title of this spec... which makes me wonder what we would re-name this one to? Perhaps `The SPIFFE Identity and Verifiable Identity Document`?

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: > Is this only one URI SAN, or only one spiffe URI SAN?

It is only one URI SAN. This decision was made to support the use of URI name constraints, though they have since been marked as optional. Do you think this limitation will impose adoption or usability pain?

> What if I wanted to have an X.509 cert that includes both the identity and the naming information?

I'm unsure what you mean by naming information... If you're asking about "regular" distinguished name validation? If so, it's assumed that folks will use DNS SAN type or oldschool DN/CN validation

> Or what about systems that have multiple identities? Is the assumption that they will use SNI to choose a cert corresponding to the requested entity?

Pretty much... if a workload has more than one SPIFFE Identity (which the current model and spec support), it is assumed that the workload can intelligently select among them. Exactly how this may be done is undefined... aka, we're not quite sure :) we may wait to see what patterns emerge before being opinionated about how this is handled (unless you have some ideas)

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: Good question :)

I propose we change it to `is the only type which may serve to identify a resource or caller`

File: standards/SPIFFE.md
Comment (smawson) [reply standards/SPIFFE.md]: Yes that sounds right to me. :+1:

File: standards/X509-SVID.md
Comment (smawson) [reply standards/X509-SVID.md]: So long as you can still use DNS SANs in addition to the SPIFFE SAN that sounds good to me.

File: standards/X509-SVID.md
Comment (smawson) [reply standards/X509-SVID.md]: :+1:

File: standards/X509-SVID.md
Comment (y2bishop2y) [reply standards/X509-SVID.md]: The one question I would have about this is what happens if we have NameConstraints and SANs with different trust domains. IE 

SAN: URI:spiffe://internal.corp.com/blah 
SAN: URI:spifffe://external.corp.com/blah

NC: URI:internal.corp.com

File: standards/X509-SVID.md
Comment (cyli) [reply standards/X509-SVID.md]: Sorry for taking so long to answer.  It's probably mostly an academic question as to whether a X.509 SVID chain is different than a regular X.509 chain.  Probably for best interop with regular mTLS certs, we shouldn't have an opinion and leave it up to the implementer of the validator service.  Seems like a very strict SPIFFE validation service would care to make sure that trust domains are SPIFFE validated all the way up to the top of chain (hence it would root its own trust at the first CA cert that has a SPIFFE ID), but it probably makes sense for the spec not to require it.

File: standards/SPIFFE-ID.md
Comment (cyli) [inline standards/SPIFFE-ID.md]: Nitpick:  "within the SPIFFE ID's trust domain" (apostrophe)

File: standards/X509-SVID.md
Comment (cyli) [inline standards/X509-SVID.md]: Non-blocking:
> in which one name-constrained trust domain issues signing certificates for a "child" trust domain.

I could just be reading this oddly, but to me this implies that we want to support hierarchical CA certs in which all the CA certs have name constraints, but I got the impression we also want to support the case where a general CA cert without name constraints can issue a signing certificates that do have name constraints and SPIFFE IDs?  And above in section 3.2, it says "A signing certificate MAY be used to issue further signing certificates in the same or different trust domains.", so wouldn't restricting this validation step to between the leaf and its immediate signing certificate preserve the case where a signing certificate chains up to one with a different trust domain?

**Review by cyli — APPROVED** [review]
Very minor grammar/phrasing nitpicks, but other than that LGTM.

File: standards/SPIFFE-ID.md
Comment (evan2645) [reply standards/SPIFFE-ID.md]: Fixed, thanks!

File: standards/X509-SVID.md
Comment (evan2645) [reply standards/X509-SVID.md]: I had included the last sentence because I think it's good to include reasoning behind certain restrictions, but in this case it seems I just made things more confusing :) I went ahead and removed the last sentence. Thanks for the review @cyli !

**Pull request merged**
